keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -402,28 +403,45 @@ class SqlQueue(
       jooq.transaction { config ->
         val txn = DSL.using(config)
 
-        txn.insertInto(messagesTable)
-          .set(idField, ulid.toString())
-          .set(fingerprintField, fingerprint)
-          .set(bodyField, mapper.writeValueAsString(message))
-          .set(updatedAtField, clock.millis())
-          .onDuplicateKeyUpdate()
-          .set(idField, MySQLDSL.values(idField) as Any)
-          .set(bodyField, MySQLDSL.values(bodyField) as Any)
-          .execute()
+        val existingFingerprint = txn.select(field(idField))
+          .from(unackedTable)
+          .where(fingerprintField.eq(fingerprint))
+          .fetchOne(0, String::class.java)
+
+        // If the fingerprint exists in the unacked table, then the message already exists. In this
+        // case, we'll reschedule the message for the delay amount, rather than adding a new
+        // message onto the queue.
+        if (existingFingerprint != null) {
+          txn.update(queueTable)
+            .set(deliveryField, atTime(delay))
+            .where(fingerprintField.eq(fingerprint))
+            .execute()
```

Comment:
This unfortunately won't work - keiko needs to support queueing a message that's on unacked, otherwise handlers like `ContinueParentStageHandler` can't requeue the message it's actively processing, and it will drop off the queue entirely.

After revisiting `RedisQueue.push()`, it's doing the following:
 - Check if the fingerprint is on the queue. 
   - If yes, only update its delivery time if it's still on the queue with set-if-exists. If it's been pulled into unacked in between the read and the update, it's a no-op and isn't requeued.
      - This prevents a narrow duplication case relevant to synthetic stage fork/join flows.
      - But still allows a handler to requeue its current message (still on unacked)..

I think the behavior in this PR is a lot better though. I think there's some duplication exposure in the redis implementation, it's just narrower because the individual query times are faster.

I think it would be great and generally safer if queue.push() disallowed pushing a message that's currently in unacked, as this PR intends. Might be worth considering adding something like an `ackAndRequeue()` method to keiko. Then switch the orca handlers that requeue their current message over to use it. Finally, ensure that `push()` won't enqueue an unacked message.

Reply:
Hmm, yeah, I can get behind that - will require some more legwork but is probably a better end solution.